File: src/binarysearch/BinarySearch.java
package binarysearch;

/*
   @author Austin Oyugi
   @since 26/11/2024
   @mail austinoyugi@gmail.com
*/
public class BinarySearch {
    public static int search(int[] nums, int target) {
        return binarySearch(nums,target,0,nums.length-1);
    }

    private static int binarySearch(int[] nums, int target,int startIndex, int endIndex){
        int arraySearchLength = startIndex + endIndex;

        int midValueIndex = arraySearchLength / 2;
        int midValue = nums[midValueIndex];
        if (target == midValue) return midValueIndex;

        if (endIndex-startIndex <= 1){
            if(nums[endIndex] == target) return endIndex;
            else return -1;
        }

        int searchStart = startIndex;
        int searchEnd = endIndex;

        if(midValue < target){
            searchStart = midValueIndex + 1;
        }else {
            searchEnd = midValueIndex -1;
        }

        return binarySearch(nums,target,searchStart,searchEnd);
    }

    public static void main(String[] args) {
        System.out.println(search(new int[]{-1,0,2,4,6,8},2));
    }
}

File: src/binarysearch/binary_search.go
package main

import "fmt"

func main() {
	nums := []int{1, 0, 2, 4, 6, 8}
	fmt.Println(search(nums, 3))
}

func search(nums []int, target int) int {
	return binarySearch(nums, target, 0, len(nums)-1)
}

func binarySearch(nums []int, target int, startIndex int, endIndex int) int {
	searchSpace := startIndex + endIndex

	midValueIndex := searchSpace / 2

	// Check if the value is in the middle index
	midValue := nums[midValueIndex]
	if midValue == target {
		return midValueIndex
	}

	//Assuming We have a 2 length array
	//We expect the previous execution to check the 0 index
	// This is the terminating condition
	if (endIndex - startIndex) <= 1 {
		if nums[endIndex] == target {
			return endIndex
		}

		// not found in the array
		return -1
	}

	// We update the search space
	newStartIndex := startIndex
	newEndIndex := endIndex

	if midValue < target {
		newStartIndex = midValueIndex + 1
	} else {
		newEndIndex = midValueIndex - 1
	}

	return binarySearch(nums, target, newStartIndex, newEndIndex)
}
